README.md
# Debate Spyder
A debate data crawler made at Data Analytics Camp 2018 at SDSU with BK, @AnyaOlson, and @anser1


curl/CHANGES.txt
                                  _   _ ____  _
                              ___| | | |  _ \| |
                             / __| | | | |_) | |
                            | (__| |_| |  _ <| |___
                             \___|\___/|_| \_\_____|

                                  Changelog

Version 7.60.0 (15 May 2018)

Daniel Stenberg (15 May 2018)
- RELEASE-NOTES: 7.60.0 release

- THANKS: added people from the curl 7.60.0 release

- docs/libcurl/index.html: removed
  
  The HTML files are long gone from the dist, now remove the last HTML
  file pointing to those missing files.
  
  d

- [steini2000 brought this change]

  http2: remove unused variable
  
  Closes #2570

- [steini2000 brought this change]

  http2: use easy handle of stream for logging

- gcc: disable picky gcc-8 function pointer warnings in two places
  
  Reported-by: Rikard Falkeborn
  Bug: #2560
  Closes #2569

- http2: use the correct function pointer typedef
  
  Fixes gcc-8 picky compiler warnings
  Reported-by: Rikard Falkeborn
  Bug: #2560
  Closes #2568

- CODE_STYLE: mention return w/o parens, but sizeof with
  
  ... and remove the github markdown syntax so that it renders better on
  the web site. Also, don't use back-ticks inlined to allow the CSS to
  highlight source code better.

- [Rikard Falkeborn brought this change]

  examples: Fix format specifiers
  
  Closes #2561

- [Rikard Falkeborn brought this change]

  tool: Fix format specifiers

- [Rikard Falkeborn brought this change]

  ntlm: Fix format specifiers

- [Rikard Falkeborn brought this change]

  tests: Fix format specifiers

- [Rikard Falkeborn brought this change]

  lib: Fix format specifiers

- contributors.sh: use "on github", not at

- http2: getsock fix for uploads
  
  When there's an upload in progress, make sure to wait for the socket to
  become writable.
  
  Detected-by: steini2000 on github
  Bug: #2520
  Closes #2567

- pingpong: fix response cache memcpy overflow
  
  Response data for a handle with a large buffer might be cached and then
  used with the "closure" handle when it has a smaller buffer and then the
  larger cache will be copied and overflow the new smaller heap based
  buffer.
  
  Reported-by: Dario Weisser
  CVE: CVE-2018-1000300
  Bug: https://curl.haxx.se/docs/adv_2018-82c2.html

- http: restore buffer pointer when bad response-line is parsed
  
  ... leaving the k->str could lead to buffer over-reads later on.
  
  CVE: CVE-2018-1000301
  Assisted-by: Max Dymond
  
  Detected by OSS-Fuzz.
  Bug: https://curl.haxx.se/docs/adv_2018-b138.html
  Bug: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=7105

Patrick Monnerat (13 May 2018)
- cookies: do not take cookie name as a parameter
  
  RFC 6265 section 4.2.1 does not set restrictions on cookie names.
  This is a follow-up to commit 7f7fcd0.
  Also explicitly check proper syntax of cookie name/value pair.
  
  New test 1155 checks that cookie names are not reserved words.
  
  Reported-By: anshnd at github
  Fixes #2564
  Closes #2566

Daniel Stenberg (12 May 2018)
- smb: reject negative file sizes
  
  Assisted-by: Max Dymond
  
  Detected by OSS-Fuzz
  Bug: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=8245

- setup_transfer: deal with both sockets being -1
  
  Detected by Coverity; CID 1435559.  Follow-up to f8d608f38d00. It would
  index the array with -1 if neither index was a socket.

- travis: add build using NSS
  
  Closes #2558

- [Sunny Purushe brought this change]

  openssl: change FILE ops to BIO ops
  
  To make builds with VS2015 work. Recent changes in VS2015 _IOB_ENTRIES
  handling is causing problems. This fix changes the OpenSSL backend code
  to use BIO functions instead of FILE I/O functions to circumvent those
  problems.
  
  Closes #2512

- travis: add a build using WolfSSL
  
  Assisted-by: Dan Fandrich
  
  Closes #2528

- RELEASE-NOTES: typo

- RELEASE-NOTES: synced

- [Daniel Gustafsson brought this change]

  URLs: fix one more http url
  
  This file wasn't included in commit 4af40b3646d3b09 which updated all
  haxx.se http urls to https. The file was committed prior to that update,
  but may have been merged after it and hence didn't get updated.
  
  Closes #2550

- github/lock: auto-lock closed issues after 90 days of inactivity

- vtls: fix missing commas
  
  follow-up to e66cca046cef

- vtls: use unified "supports" bitfield member in backends
  
  ... instead of previous separate struct fields, to make it easier to
  extend and change individual backends without having to modify them all.
  
  closes #2547

- transfer: don't unset writesockfd on setup of multiplexed conns
  
  Curl_setup_transfer() can be called to setup a new individual transfer
  over a multiplexed connection so it shouldn't unset writesockfd.
  
  Bug: #2520
  Closes #2549

- [Frank Gevaerts brought this change]

  configure: put CURLDEBUG and DEBUGBUILD in lib/curl_config.h
  
  They are removed from the compiler flags.
  
  This ensures that make dependency tracking will force a rebuild whenever
  configure --enable-debug or --enable-curldebug changes.
  
  Closes #2548

- http: don't set the "rewind" flag when not uploading anything
  
  It triggers an assert.
  
  Detected by OSS-Fuzz
  Bug: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=8144
  Closes #2546

- travis: add an mbedtls build
  
  Closes #2531

- configure: only check for CA bundle for file-using SSL backends
  
  When only building with SSL backends that don't use the CA bundle file
  (by default), skip the check.
  
  Fixes #2543
  Fixes #2180
  Closes #2545

- ssh-libssh.c: fix left shift compiler warning
  
  ssh-libssh.c:2429:21: warning: result of '1 << 31' requires 33 bits to

nsda_test.R
source("~/R/debate-spyder/nsda.R")

library(xtable)
# log = NSDAlogin("nepalsamyok@gmail.com", "AbeLincoln", "cookies/3")
s = NSDASearch("Bidhi Kasu", cookies = "cookies/1")
x = NSDAPoints(s[1,1],cookies="cookies/1")

NSDAthreatScore(x)

Group Project/nsda_test.R

username = "nepalsamyok@gmail.com"
password = "AbeLincoln" # for NSDA
loginName=URLencode(username)
pwd = URLencode(password)

curl = "C:/Users/SSO103/Downloads/curl-7.60.0-win32-mingw/bin/curl -s " 

base_url= '"https://api.speechanddebate.org/v1/'

thing = system(paste('C:/Users/SSO103/Downloads/curl-7.60.0-win32-mingw/bin/curl -s -c nsda.cookies -d "log=',loginName,'&pwd=',password,'&wp-submit=Log+In&redirect_to=%2Faccount&testcookie=1" -X POST https://www.speechanddebate.org/wp-login.php'))

person_to_search = "Samyok Nepal"

person = URLencode(person_to_search)
search = (paste(curl, base_url, 'search?q=', person, '" -b nsda.cookies', sep=''))
results = system(search, intern = TRUE)
#thingC:/Users/SSO103/Downloads/curl-7.60.0-win32-mingw/bin>curl 
library(jsonlite)
mydf <- fromJSON(results)
mydf[1,1] # id of first result

app.R
library(shiny)
library(xtable)
source("nsda.R")
# Define server logic ----
server <- function(input, output, session) {
  logged_in <- FALSE
  output$login_errors <- renderText({
    "Login Please."
  })
  output$log_text <- renderText({
    "Login"
  })
  ## LOGIN ========================================================M
  observeEvent(input$login_btn, {
    #
    output$login_errors <- renderText({
      #
      "Loading..."                                                 #
    })                                                             #
    if (NSDAlogin(input$uname,
                  #
                  input$password,
                  #
                  concat('cookies/', session$token))) {
      #
      logged_in = TRUE                                             #
      output$log_text <- renderText({
        #
        if (logged_in) {
          #
          "Logout"
        } else {
          "Login"
        }
      })
      output$login_errors <- renderText({
        #
        "Logged In!"                                               #
      })                                                           #
    } else {
      #
      output$login_errors <- renderText({
        #
        "Login Failed!"                                            #
      })                                                           #
    }                                                              #
  })                                                               #
  ## /LOGIN =======================================================W
  
  observeEvent(input$individualsearchbutton, {
    x = NSDAMember(
      NSDASearch(
        input$individualsearch,
        cookies =  concat('cookies/', session$token)
      )[1, 1],
      concat('cookies/', session$token)
    )
    print(x)
    output$individuals <- renderTable(x)
  })
  observeEvent(input$competitionbtn, {
    ppl = as.list(strsplit(input$competitionsearch, "\\n")[[1]])
    final = data.frame()
    for(i in 1:length(ppl)){
      peopleWithSameName =NSDASearch(ppl[[i]][1])
      for(j in 1:nrow(peopleWithSameName)){
        member = NSDAMember(peopleWithSameName[j, 'id'], cookies= "cookies/1")
        x = NSDAPoints(peopleWithSameName[j,1],cookies="cookies/1")
        
        new = c("threat level"=NSDAthreatScore(x = x, member = member))
        rbind(final, new)
      }
    }
    output$competition <- renderDataTable(final)
  })
}

ui <- navbarPage(
  id = "home",
  "Debate SPYDER",
  tabPanel("Home", fillPage(
    h1(span("Debate SPYDER", style = "color:#d00000"),
       align = "center"),
    p(
      "A website dedicated to helping you",
      span("analyze" , style = "color:#d00000 "),
      "and",
      span("interpret", style = "color:#d00000"),
      "debate data",
      align = "center",
      style = "font-size:25px"
    ),
    p(
      "If you know the opponent and know yourself, you need not fear the result of a hundred debate rounds. - Sun Tzu, modified",
      align = "center",
      style = "font-size:15px"
    ),
    img(src = "http://cdn.samyok.us/SPYDER.png",
        
        style = "display: block; margin-left: auto; margin-right: auto; width=100%"),
    h4("Credits:")
  )),
  navbarMenu(
    "Search",
    tabPanel(
      "Schools",
      textInput("schoolsearch", label = "School Search", placeholder = "Search here..."),
      actionButton("schoolsearchbutton", label = "Go"),
      tableOutput('schools')
    ),
    tabPanel(
      "Competition",
      selectInput(
        "competitionType",
        h3("Debate Type"),
        choices = list(
          "Lincoln-Douglas Debate" = 102,
          "Policy Debate" = 104,
          "Public Forum Debate" = 103
        ),
        selected = 103
      ),
      textAreaInput(
        "competitionsearch",
        label = "Competition Search",
        height = "300px",
        width = "900px"
      ),
      actionButton("competitionbtn", label = "Go"),
      dataTableOutput('competition')
      ),
      tabPanel(
        "Threats",
        textInput("threatsearch", h3("Name")),
        selectInput(
          "selectdebate",
          h3("Debate Type"),
          choices = list(
            "Lincoln-Douglas Debate" = 102,
            "Policy Debate" = 104,
            "Public Forum Debate" = 103
          ),
          selected = 103
        ),
        actionButton("threatsearchbutton", label = "Go")
      ),
      tabPanel(
        "Individuals",
        textInput("individualsearch", label = "Individual Search", placeholder = "Search here..."),
        actionButton("individualsearchbutton", label = "Go"),
        tableOutput('individuals')
      )
    ),
    tabPanel(
      "Info",
      h1("Information"),
      h2("Search"),
      h3(span("Schools", style = "color:#d00000")),
      p("View any school's details and statistics.", style = "font-size:20px"),
      h3(span("Competition", style = "color:#d00000")),
      p(
        "Copy and paste the list of your competitors to analyze all of their statistics at once, including state, district, school, and points.",
        style = "font-size:20px"
      ),
      h3(span("Individuals", style = "color:#d00000")),
      p(
        "Search for any individual and view details and statistics.",
        style = "font-size:20px"
      ),
      h2("Threat Algorithm"),
      p(
        "To view someone's threat rating, search up his/her name in the Threats page and specify the debate type.
        We can give a threat ranking in each of the three main types of debate: Lincoln-Douglas Debate, Policy Debate, and Public Forum Debate.",
        style = "font-size:20px"
      ),
      p(
        "Our threat algorithm incorporates statistics including but not limited to points in the debate category,
        number of years of participation, win/loss ratio, and number of Nationals points.",
        style = "font-size:20px"
      ),
      p(
        "Increasing your threat rating all comes down to putting more time, effort, and practice into debate.",
        style = "font-size:20px"
      ),
      p(
        "Keep in mind that this algorithm's output is not perfect.
        Just because your rating is higher or lower than another person does",
        strong("not"),
        "mean that you will definitely win or lose, respectively.
        This is because there is always an element of luck and chance in debate rounds, and our algorithm, like everything else, can not predict winners.
        The rating's purpose is to give you a rough estimate of your opponent's level and for entertainment.
        How seriously you take it is entirely up to you, just like how seriously you take NSDA points.",
        style = "font-size:20px"
      ),
      p(
        "Finally, never underestimate or overestimate your opponents. Good luck!",
        style = "font-size:20px"
      ),
      h2("Topics"),
      p(
        "In this page, we will list the current Big Questions, Lincoln-Douglas, Policy, and Public Forum resolutions when available.
        When the Nationals topics come out, we will also list the World Schools Debate resolutions.",
        style = "font-size:20px"
      ),
      h2("Login"),
      p(
        "Login using your National Speech and Debate Association credentials.
        We do not collect your login information.",
        style = "font-size:20px"
      )
    ),
    tabPanel(
      textOutput("log_text"),
      inline = TRUE,
      textInput("uname", "Username:"),
      passwordInput("password", "Password:"),
      actionButton("login_btn", "Go"),
      h3(textOutput("login_errors"), style = "color:red")
    ),
    inverse = TRUE,
    collapsible = TRUE,
    includeCSS("http://cdn.samyok.us/app.css")
      )
  options("shiny.port" = 8000)
  # Run the app ----
  shinyApp(ui = ui,
           server = server)
  

nsda.R
# lib for samyok
library(jsonlite)
library(readr)

CURL_DIR = paste('curl', sep = '')
concat = function(...) {
  return(paste(..., sep = ''))
} ## WORKS
SCurl = function(url, POST = FALSE, data = NULL, cookies = "nsda.cookies") {
  if (POST) {
    rslt = system(concat(CURL_DIR, ' -s -b ', cookies, ' -d ', data, ' -X POST ', url),
                  intern = TRUE)
    print(s)
  } else {
    rslt = system(concat(CURL_DIR, ' -s -b ', cookies, ' ', url), intern = TRUE)
  }
  return(rslt)
} ## WORKS
NSDAlogin = function(user, pass, cookies) { #### WORKS
  loginName = URLencode(user, reserved =TRUE)
  pwd = URLencode(pass, reserved =TRUE)
  thing = system(
    concat(
      'curl -s -c "',
      cookies,
      '" -d "log=',
      loginName,
      '&pwd=',
      pwd,
      '&wp-submit=Log+In&redirect_to=%2Faccount&testcookie=1" -X POST https://www.speechanddebate.org/wp-login.php'
    ))
  test_search = fromJSON(SCurl(url="https://api.speechanddebate.org/v1/search?q=0", cookies = cookies))
  return(is.data.frame(test_search))
} ## WORKS

## EVERYTHING ABOVE WORKS
## SAMYOK 6/20/18 21:23 CST

NSDASearch = function(name, cookies = "nsda.cookies", type = "member" ) {
  nam = URLencode(name, reserved = FALSE)  
  results = fromJSON(SCurl(concat("https://api.speechanddebate.org/v1/search?q=",nam), cookies, POST = FALSE))
  final = data.frame("id"=1, "name"="[[NO RESULTS]]", "type"='member', stringsAsFactors = FALSE)
  index = 1
  for(i in 1:length(results[,1])){
    if(results[i,]$type == type){
      final[index, ] = results[i,]
      index <- index + 1 
    }
  }
  return(final)
}
NSDAMember <-  function(id, cookies="nsda.cookies") {
  x = fromJSON(SCurl(concat("https://api.speechanddebate.org/v1/members/",id), cookies, POST = FALSE))
  backlist = c("degrees", "honors", "citations", "timestamp")
  for(i in 1:length(backlist)){
    x[backlist[i]] = NULL
  }
  # xtable(data.frame((x)))
  
  lname <- names(x)
  y <- data.frame()
  # names(y) <- lname
  z <- c()
  for(i in 1:length(lname)){
    if(is.null(x[[i]] )){
      z[i] <- ""
    } else {
      z[i] <- (x[[i]])
    }
    
  }
  
  mx = matrix(z, nrow = 1, ncol = 22)
  df = as.data.frame(mx)
  names(df) <- lname
  return(df)
}
NSDAPoints <- function(id, cookies = "nsda.cookies"){
  x = fromJSON(SCurl(concat("https://api.speechanddebate.org/v1/members/",id, "/points"), cookies, POST = FALSE))
  return(x)
}
NSDAGetCompPoints <- function(arr, competition_type){
  z = data.frame()
  
  for(i in 1:nrow(x)){
    if(x[i,'category'] == competition_type){
      z = rbind(z, x[i,])
    }
  }
  return(z)
}
NSDANumWins = function(y){
  result = 0
  for(i in 1:nrow(y)){
    result = result + as.numeric(strsplit(y[i, 'result'], "W")[[1]][1])
  }
  return(result)
}
NSDANumLosses = function(y){
  result = 0
  for(i in 1:nrow(y)){
    result = result + as.numeric(strsplit(strsplit(y[i, 'result'], "W, ")[[1]][2], "L")[[1]][1])
  }
  return(result)
}

NSDAEarliestYear = function(y) {
  result = 999999
  for(i in 1:nrow(y)){
    year = as.numeric(strsplit(y[i, 'tourn_start'], "-")[[1]][1])
    if( year < result) {
      result = year  
    }
  }
  return(result)
}
NSDANumNats  =function(y){
  result = 0
  for(i in 1:nrow(y)){
    if(y[i, 'nationals']){
      result = result +1 
    }
  }
  return(result)
}
NSDANatsPts = function(y){
  result = 0
  for(i in 1:nrow(y)){
    if(y[i, 'nationals']){
      result = result + y[i, 'points']
    }
  }
  return(result)
}
NSDAthreatScore = function(x, member){
  y = NSDAGetCompPoints(x, 103)
  points = sum(y['points'])
  
  summy = NSDANumWins(y)
  lossy = NSDANumLosses(y)
  
  years = NSDAEarliestYear(y) - as.numeric(format(Sys.Date(), "%Y"))
  nationals = NSDANumNats(y)
  ptsNats = NSDANatsPts(y)
  
  if(is.null(member[1, 'points_last_year'])){
    ptsLastYear = 0
  } else {
    ptsLastYear =as.numeric(as.character( member[1, 'points_last_year']))
  }
  ptsThisYear = as.numeric(as.character(member[1, 'points_this_year']))
  totalPoints = as.numeric(as.character(as.numeric(member[1, 6])))
  
  threat_score = points/10 + totalPoints/20+ ptsThisYear/10 + years*50 + summy*2 + summy * 20 / (lossy +1 ) + nrow(x) + nationals*50 + ptsNats/5
}
